lib/ecto/changeset.ex

```diff
@@ -1015,7 +1024,22 @@ defmodule Ecto.Changeset do
           "please use put_#{tag}/4 instead"
   end
 
-  defp put_change(data, changes, errors, valid?, key, value, _type) do
+  defp put_change(data, changes, errors, valid?, key, value, nil) do
+    put_change(data, changes, errors, valid?, key, value)
+  end
+
+  defp put_change(data, changes, errors, valid?, key, value, type) do
+    case Ecto.Type.cast(type, value) do
```

Comment:
looking at the clauses below we could use `Ecto.Type.dump` instead, because the given value must exactly match type type. WDYT?